File: leetcode/recursion + memo/282. Expression Add Operators/py/main.py
class Solution:
    # Time O(4^N)
    # Space O(N)
    def addOperators(self, num, target):
        n = len(num)
        ans = []

        def dfs(index, prevNumber, curNumber, totalSum, items):
            # Done processing all the digits in num
            if index >= n:
                # If the final value == target expected AND no unprocessed numbers
                if totalSum == target and curNumber == 0:
                    ans.append(''.join(items)[1:])
                return

            # Extending the current operand by one digit
            curNumber = curNumber * 10 + int(num[index])
            strCurNumber = str(curNumber)

            # To avoid cases where we have 1 + 05 or 1 * 05 since 05 won't be a valid operand.
            # Extend current number
            if curNumber > 0:
                dfs(index + 1, prevNumber, curNumber, totalSum, items)

            # Apply operations to the current number

            items.append('+')
            items.append(strCurNumber)

            dfs(index + 1, curNumber, 0, totalSum + curNumber, items)

            items.pop()
            items.pop()

            # Can subtract or multiply only if there are some previous operands
            if items:
                # SUBTRACTION
                items.append('-')
                items.append(strCurNumber)

                dfs(index + 1, -curNumber, 0, totalSum - curNumber, items)

                items.pop()
                items.pop()

                # MULTIPLICATION
                items.append('*')
                items.append(strCurNumber)

                dfs(index + 1, prevNumber * curNumber, 0,
                    totalSum - prevNumber + (curNumber * prevNumber), items)

                items.pop()
                items.pop()

        dfs(0, 0, 0, 0, [])

        return ans

File: leetcode/recursion + memo/282. Expression Add Operators/js/main.js
// Time O(N * 4^N)
// Space O(N)
const addOperators = (num, target) => {
  let n = num.length;
  let ans = [];

  dfs(0, 0, 0, 0, []);

  return ans;

  function dfs(index, prevOper, curOper, value, string = []) {
    if (index >= n) {
      if (value == target && curOper == 0) {
        ans.push(string.join('').substring(1));
      }
      return;
    }

    curOper = curOper * 10 + parseInt(num[index]);

    if (curOper > 0) {
      dfs(index + 1, prevOper, curOper, value, string);
    }

    string.push('+');
    string.push(curOper);
    dfs(index + 1, curOper, 0, value + curOper, string);
    string.pop();
    string.pop();

    if (string.length) {
      string.push('-');
      string.push(curOper);
      dfs(index + 1, -curOper, 0, value - curOper, string);
      string.pop();
      string.pop();

      string.push('*');
      string.push(curOper);
      dfs(index + 1, curOper * prevOper, 0, value - prevOper + curOper * prevOper, string);
      string.pop();
      string.pop();
    }
  }
};
